### ISSUE TITLE
Question About Gradient Calculation Method

### ISSUE THREAD

**Issue by hellojxt** [LINs-lab/UCGM#3]
Hi, thank you for your excellent work! I have a question: why did you choose to compute the gradient of $f_t^x$ (Equation 6 in the paper) using numerical methods instead of PyTorch’s built-in automatic differentiation tools, such as the `jvp` function mentioned in [[this paper](https://arxiv.org/pdf/2505.13447)]? In theory, these autodiff-based methods should be more stable and accurate than numerical differentiation. I'm wondering if there's a specific reason for this choice. Also, as a follow-up: when computing gradients with respect to a frozen model, would using the EMA version lead to more stable results? I noticed that your code seems to use the current model parameters instead of the EMA version—was that a deliberate choice?

**Comment by sp12138** [comment]
Thank you for your question. In fact, using the second-order finite difference method (numerical methods) to estimate derivatives is the optimal solution I arrived at after investigating all applicable mathematical/computer theories, for the following reasons:  

(1) Compared to the second-order finite difference method, using JVP (Jacobian-vector product) to compute derivatives leads to higher memory demands and increased computational latency.  
(2) Flash Attention, commonly used in Diffusion Transformers, is incompatible with JVP. This means using JVP would require abandoning Flash Attention, further reducing computational efficiency.  
(3) Given the incompatibility of JVP with technologies like Flash Attention, its use could diminish the framework's compatibility. This contradicts the topic of our paper (i.e., "unified generative models"), as we aim to design a highly compatible and user-friendly framework.  
(4) In practice, the second-order finite difference method has proven sufficient for training few-step models in all our experimental scenarios, despite its theoretical gap compared to JVP.  

As a result, our framework achieves superior performance. _**For example, UCGM takes only 8 minutes (with 8 GPUs) to fine-tune a pre-trained diffusion model on ImageNet-1K 256x256 into a few-step model (NFE=2, FID=1.39)**._ This result far surpasses the best outcomes mentioned in [the paper you cited](https://arxiv.org/pdf/2505.13447), which required extensive computational resources for training.

You can check more details about our UCGM-tuned few-step models in [README.md](https://github.com/LINs-Lab/UCGM).

**Comment by sp12138** [comment]
Regarding your second question, i.e., whether the parameters $\theta$ of EMA can be used to compute directional derivatives, I believe this is not feasible. The reason is that we specifically need to calculate the directional derivatives with respect to the current parameters $\theta$. Although using EMA might intuitively provide a more stable estimation of directional derivatives, I think this approach lacks mathematical justification and may not be a good solution.  

(Note: EMA stands for Exponential Moving Average, which is often used in optimization algorithms to smooth parameter updates.)

**Comment by sp12138** [comment]
@hellojxt 
At last, thank you very much for your question, which inspired me to add more discussion on how to effectively compute directional derivatives in the paper. If you have any other questions, feel free to ask further.

**Comment by hellojxt** [comment]
<img width="1110" alt="Image" src="https://github.com/user-attachments/assets/cae4fcd9-0f26-4edd-b6fb-01bcc237a8ef" />

Thank you very much for your detailed response! As you mentioned, UCGM can quickly fine-tune a multi-step model into a highly effective few-step model, which is truly exciting and encouraging.

Regarding the fine-tuning strategy shown in the screenshot, I also wanted to say that using the pretrained model as guidance is a elegant and efficient idea! I have a couple of follow-up questions. Why is $\zeta$ fixed to 1/2 when $t > s$? Also, when training the few-step model, are the weights initialized from the pretrained multi-step model? In addition, based on the table in the README, some few-step models are said to be fine-tuned from pretrained models (e.g., REPA-E). Since those pretrained models were not trained with guidance, does that mean the guidance mechanism is essentially inactive during fine-tuning? Or were the pretrained models in the README actually trained using UCGM-T?

I’d be very grateful if you could share your thoughts on these questions as well.

**Comment by sp12138** [comment]
@hellojxt 
(1) The choice of ζ = 1/2 when t > s is an empirical approach inspired by [Paper](https://arxiv.org/abs/2404.07724), which demonstrated that guidance should only be applied during a specific time interval. We generalize this key insight by proposing that different guidance strengths should be applied to different intervals, hence our use of varying ζ values across different timesteps t.

(2) For few-step model training, we initialize the weights from the pretrained multi-step model. This implementation detail is available in our provided code.

(3) UCGM can also convert conventionally trained models (e.g., REPA-E) into few-step versions through tuning. As noted in our [News section](https://github.com/LINs-Lab/UCGM?tab=readme-ov-file#fire-news), the supporting code and configuration files for this tuning capability will be released shortly—the current version does not yet include this functionality.

In summary, UCGM provides three key capabilities:

1. UCGM-S: Training-free, plug-and-play sampling acceleration
2. UCGM-T: Unified training framework for both few-step and multi-step models
3. UCGM-T: Ultra-efficient tuning of conventionally trained models (e.g., REPA-E) into few-step versions (Code will be coming soon)

**Comment by sp12138** [comment]
Hello @hellojxt ,
The tuning code has been uploaded. If everything looks good on your end, I'll go ahead and close this issue.

Best wishes to you.

**Comment by hellojxt** [comment]
Thank you for providing the tuning code—it’s been very helpful! I don’t have any further questions, so I’ll go ahead and close this issue.

Best wishes!
